environment.go

```diff
@@ -31,5 +33,46 @@ func (r *EnvironmentVariableResolver) Resolve(keys []string, _ ConfigMap) (val i
 	if len(r.Prefix) > 0 {
 		keysCopy = append([]string{r.Prefix}, keysCopy...)
 	}
-	return os.LookupEnv(strings.Join(toUpperStringSlice(keysCopy), "_"))
+
+	var translator KeyTranslator
+	if r.Translator == nil {
+		// If we weren't given a specific translator to use, then use the
+		// default translator.
+		translator = DefaultEnvironmentVariableKeyTranslator
+	} else {
+		// Otherwise, use the translator that was provided when this resolver
+		// was created.
+		translator = r.Translator
+	}
+
+	return os.LookupEnv(toEnvironmentVariable(keysCopy, translator))
+}
+
+// The DefaultEnvironmentVariableKeyTranslator is the default KeyTranslator
+// used by the EnvironmentVariableResolver.
+//
+// It is responsible for mapping arbitrary input keys to an environment
+// variable to lookup. This is done by converting all characters to upper case
+// and replacing any hyphens with underscores.
+func DefaultEnvironmentVariableKeyTranslator(b byte) byte {
+	switch b {
+	case '-':
+		return '_'
+	default:
+		return byte(unicode.ToUpper(rune(b)))
+	}
+}
+
+func toEnvironmentVariable(keys []string, translator KeyTranslator) string {
+	// Convert the input keys into a single environment variable that we can
+	// perform a lookup on.
+	key := []byte(strings.Join(keys, EnvSeparator))
+
+	// Next, perform any custom translations performed by the KeyTranslator.
+	for index, char := range key {
+		key[index] = translator(char)
```

Comment:
Given you're already importing `strings` from std lib, why not just work with `string` type instead of having to convert to byte slices merely because that's how you've defined the interface. By that I mean, the default behavior you're introducing is to replace `-` with `_` in the key (which is already a string) so maybe just make the interface `KeyTranslator(string) string` and use the std lib `strings.ReplaceAll`?

Reply:
I mostly went with this API because I _didn't_ want to provide the ability for a translator to wholesale rewrite some/all of the key and instead to only allow it to translate some subset of characters if needed, but the individual units of a string are `byte`s so I think the API needs to stay to ensure we don't give Translators too much control over the keys